src/k4a_ros_device.cpp

```diff
@@ -854,28 +984,93 @@ void K4AROSDevice::imuPublisherThread()
+                return;
+            }
 
-        ImuPtr imu_msg(new Imu);
+            ImuPtr imu_msg(new Imu);
 
-        result = getImuFrame(sample, imu_msg);
-        ROS_ASSERT_MSG(result == K4A_RESULT_SUCCEEDED, "Failed to get IMU frame");
+            result = getImuFrame(sample, imu_msg);
+            ROS_ASSERT_MSG(result == K4A_RESULT_SUCCEEDED, "Failed to get IMU frame");
 
-        imu_orientation_publisher_.publish(imu_msg);
+            imu_orientation_publisher_.publish(imu_msg);
+        }
+        else if (k4a_playback_handle_)
+        {
+            // publish imu messages as long as the imu timestamp is less than the last capture timestamp to catch up to the cameras
+            // compare signed with unsigned shouldn't cause a problem because timestamps should always be positive
+            while (last_imu_time_usec_ <= last_capture_time_usec_ && !imu_stream_end_of_file_)
+            {
+                std::lock_guard<std::mutex> guard(k4a_playback_handle_mutex_);
+                k4a_stream_result_t stream_result = k4a_playback_get_next_imu_sample(k4a_playback_handle_, &sample);
+                if (stream_result == K4A_STREAM_RESULT_EOF)
+                {
+                    imu_stream_end_of_file_ = true;
+                }
+                else if (stream_result == K4A_STREAM_RESULT_SUCCEEDED)
+                {
+                    ImuPtr imu_msg(new Imu);
+                    result = getImuFrame(sample, imu_msg);
+                    ROS_ASSERT_MSG(result == K4A_RESULT_SUCCEEDED, "Failed to get IMU frame");
+
+                    imu_orientation_publisher_.publish(imu_msg);
+
+                    last_imu_time_usec_ = sample.acc_timestamp_usec;
+                }
+                else
+                {
+                    ROS_FATAL("Failed to get next imu sample from recording. node cannot continue.");
+                    ros::requestShutdown();
+                    return;
+                }
+                
+            }
+        }
 
         ros::spinOnce();
         loop_rate.sleep();
     }
 }
 
+std::chrono::microseconds K4AROSDevice::getCaptureTimestamp(const k4a::capture &capture)
+{
+    // Captures don't actually have timestamps, images do, so we have to look at all the images
+    // associated with the capture.  We only need an approximate timestamp for seeking, so we just
+    // return the first one we get back (we don't have to care if a capture has multiple images but
+    // the timestamps are slightly off).
+    //
+    // We check the IR capture instead of the depth capture because if the depth camera is started
+    // in passive IR mode, it only has an IR image (i.e. no depth image), but there is no mode
+    // where a capture will have a depth image but not an IR image.
+    //
+    const auto irImage = capture.get_ir_image();
+    if (irImage != nullptr)
+    {
+        return irImage.get_device_timestamp();
+    }
+
+    const auto depthImage = capture.get_depth_image();
```

Comment:
I think on the Azure Kinect, IR and Depth timestamps will always be the same (since they're the same sensor, just different processing), so you probably don't need to have a case for checking the depth image.

Regardless, like you've said in the comment above, there's no possible way to have depth image valid and IR image invalid. Why bother with this extra case?